File: 2021/2021W17/PreppindataW17.py
import pandas as pd
import re

xlsx = pd.ExcelFile("F:/Data/Preppin Data Challenge.xlsx")

df = pd.read_excel(xlsx,sheet_name=0)
df = df.dropna(subset=[df.columns[0]]).drop([t for t in df.columns if re.match('^[Unnamed|Project]',str(t))],axis=1)     
df[df.columns[-1]] = df[df.columns[-1]].apply(lambda x: None if isinstance(x,str) else x)
df = df.melt(id_vars=[df.columns[0]], var_name='Date', value_name='Hours').dropna()
df['Name'] = df[df.columns[0]].apply(lambda x: re.search('(.*),',x).group(1))
df['Area of Work'] = df[df.columns[0]].apply(lambda x: re.search('(?<=:\s)(.*)$',x).group(1))

dk = df.groupby(['Name']).agg({'Date':'nunique', 'Hours':'sum'})
dk['Avg Hours'] = dk['Hours']/dk['Date']
final = df[df['Area of Work']!='Chats'].groupby(['Name', 'Area of Work']).agg({'Hours':'sum'})
final['% of Total'] = final.groupby(level=0).apply(lambda x: round(100 * x / float(x.sum())))
final['% of Total'] = final['% of Total'].apply(lambda x: f'{int(x)}%') 
final = final.reset_index(drop=False).drop('Hours', axis=1)
final = final[final['Area of Work']=='Client']
final['Avg Number of Hours worked per day'] = final['Name'].apply(lambda x: dk.loc[x,'Avg Hours'])

File: 2021/2021W17/preppindataW17.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)

final <- read_excel("F:/Data/Preppin Data Challenge.xlsx") %>%
  mutate_at(ncol(.),  as.numeric) %>%
  drop_na(`Name, Age, Area of Work`) %>%
  select(-'Project') %>%
  pivot_longer(cols=-'Name, Age, Area of Work', names_to='DateInNum', values_to='Hours', values_drop_na=TRUE) %>%
  mutate('Name' = str_extract(`Name, Age, Area of Work`,'^([^,]*)'),
         'Area of Work' = str_extract(`Name, Age, Area of Work`,'(?<=:\\s)(.*)$'))%>%
  group_by(`Name`) %>%
  mutate('Avg Number of Hours worked per day' = sum(`Hours`)/n_distinct(`DateInNum`),
         '% of Total' = sum(if_else(`Area of Work`=='Client',`Hours`,0))/sum(if_else(`Area of Work`=='Chats',0,`Hours`)),
         '% of Total' = paste0(round(`% of Total`*100),'%')) %>%
  filter(`Area of Work` == 'Client') %>%
  select(c('Name', 'Area of Work', '% of Total', 'Avg Number of Hours worked per day')) %>%
  distinct()

View(final)
